### PR TITLE
feat(subnets): Add support for reserved IPv6 addresses MAASENG-3162

### PR DESCRIPTION
**Pull request by ndv99**
## Done
- Implemented new designs for IPv6 in the Add/Edit forms
- Added validation for IPv6 (with help from `is-ip` and `ipaddr.js`)
- Updated `delete` action to include the IP in the payload to match backend expectations
- Re-ordered some of the code in `ReserveDHCPLease` to be a little more sensible

## QA steps

Ping me privately and I'll give you access to my container with the backend. Also, make sure the `VITE_APP_STATIC_IPS_ENABLED` is set to `true`.

- [ ] Go to `/subnet/2/address-reservation` (this is an IPv6 subnet)
- [ ] Click on "Reserve static DHCP lease"
- [ ] Ensure the routing prefix and subnet ID are displayed correctly
- [ ] Ensure the placeholder is correct (should be four sets of "0000" separated by colons
- [ ] Enter some text that definitely isn't an IPv6 interface ID
- [ ] Ensure an error message is displayed
- [ ] Remove that text
- [ ] Ensure an error message "IP address is required" is displayed
- [ ] Enter a valid IPv6 interface ID (e.g. `:1` or `FFFF:FFFF:FFFF:FFFF`
- [ ] Ensure no error messages are displayed and that the form can be submitted
- [ ] Open the "Edit" form for the address you just reserved
- [ ] Ensure the form pre-fills the address correctly

## Fixes

Fixes [MAASENG-3162](https://warthogs.atlassian.net/browse/MAASENG-3162)

## Screenshots

image

## Notes

[Figma design reference](https://www.figma.com/design/bX9XqdURp0pOX3ujuD5WpB/Reserved-IPs?node-id=202-150455&t=tsz4E3mKEfrLbdzF-1)

[MAASENG-3162]: https://warthogs.atlassian.net/browse/MAASENG-3162?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/app/subnets/views/SubnetDetails/StaticDHCPLease/ReserveDHCPLease/ReserveDHCPLease.tsx [hunk 1]
```diff
@@ -77,65 +94,56 @@ const ReserveDHCPLease = ({
...
+        },
       }),
     mac_address: Yup.string().matches(MAC_ADDRESS_REGEX, "Invalid MAC address"),
     comment: Yup.string(),
   });
 
   const handleSubmit = (values: FormValues) => {
+    const ip = subnetIsIpv4
+      ? `${immutableOctets}.${values.ip_address}`
+      : `${ipv6Prefix}${values.ip_address}`;
```
Comment (petermakowski) [inline src/app/subnets/views/SubnetDetails/StaticDHCPLease/ReserveDHCPLease/ReserveDHCPLease.tsx:140]: The handleSubmit function could be simplified, e.g. by creating a helper function to format the IP address based on the type (IPv4 or IPv6).

File: src/app/subnets/views/SubnetDetails/StaticDHCPLease/ReserveDHCPLease/ReserveDHCPLease.tsx [hunk 2]
```diff
@@ -77,65 +94,56 @@ const ReserveDHCPLease = ({
...
+          } else {
+            try {
+              const addr = ipaddr.parse(`${ipv6Prefix}${ip_address}`);
+              const netAddr = ipaddr.parse(networkAddress);
+              return addr.match(netAddr, prefixLength);
+            } catch (e) {
+              return false;
+            }
+          }
+        },
```
Comment (petermakowski) [inline src/app/subnets/views/SubnetDetails/StaticDHCPLease/ReserveDHCPLease/ReserveDHCPLease.tsx:131]: The form validation logic is quite complex and could be simplified.

File: src/app/subnets/views/SubnetDetails/StaticDHCPLease/ReserveDHCPLease/ReserveDHCPLease.test.tsx [hunk 3]
```diff
@@ -170,6 +170,38 @@ it("pre-fills the form if a reserved IP's ID is present", async () => {
   );
 });
 
+it("pre-fills the form if a reserved IPv6 address's ID is present", async () => {
```
Comment (petermakowski) [inline src/app/subnets/views/SubnetDetails/StaticDHCPLease/ReserveDHCPLease/ReserveDHCPLease.test.tsx:173]: Good job on these tests, very thorough.

File: src/app/base/components/PrefixedIpInput/PrefixedIpInput.tsx [hunk 4]
```diff
@@ -18,46 +21,76 @@ type Props = Omit<
...
+        return 7; // 6 digits, 1 dot
+      } else if (immutableOctetsLength === 1) {
+        return 11; // 9 digits, 2 dots
+      } else {
+        return 15; // 12 digits, 3 dots
+      }
+    } else {
+      return getPlaceholderText().length;
+    }
+  };
```
Comment (petermakowski) [inline src/app/base/components/PrefixedIpInput/PrefixedIpInput.tsx:62]: I'd suggest refactoring to simplify the logic.

```tsx
  const getIPv6Placeholder = () => {
    // 7 is the maximum number of colons in an IPv6 address
    const placeholderColons = 7 - (ipv6Prefix.match(/:/g) || []).length;
    return `${"0000:".repeat(placeholderColons)}0000`;
  };

  const getPlaceholderText = () =>
    subnetIsIpv4 ? editable : getIPv6Placeholder();

  const getIPv4MaxLength = () => {
    const immutableOctetsLength = immutable.split(".").length;
    const lengths = [15, 11, 7, 3]; // Corresponding to 0-3 immutable octets
    return lengths[immutableOctetsLength];
  };

  const getMaxLength = () =>
    subnetIsIpv4 ? getIPv4MaxLength() : getIPv6Placeholder().length;
```

- Clearly separates the logic for IPv4 and IPv6, making it easier to follow.
- Uses an array for IPv4 lengths, which simplifies the logic.

Using data structures for lookup is _generally a good idea_, next time when you're dealing with switch-like statements, consider using this pattern.

**Comment by ndv99** [comment]
> LGTM, one thing I noticed is that "Delete static DHCP lease" does not close on success.

Yeah, I noticed that too. Will investigate :+1:

**Review by petermakowski — APPROVED** [review]

**Pull request merged**